### ISSUE TITLE
running tests in parallel

### ISSUE THREAD

**Issue by cyphar** [bats-core/bats-core#171]
Currently tests are all run in serial, which results in suboptimal test running performance (in the case of `umoci` we've started to hit against the 50 minute limit that Travis has because we run ~200 tests that all take a fairly long amount of time -- but are easily parallelisable).

There was an issue about this in sstephenson/bats#143, but I think the easiest solution would be to use GNU parallel if it's available (which knows how to collate data between different programs and can also do a bunch of other clever things when running). In theory the fix should be quite trivial (a minor rewrite of `bats-exec-suite` to use GNU parallel for running `bats-exec-test`).

**Comment by cyphar** [comment]
Sorry, it would be a change to `bats-exec-test`'s `bats_perform_tests` such that when it re-execs itself in order to do `bats_perform_test` it's done as a GNU parallel script. The only problem is that the numbering _could_ be in the wrong order -- though GNU parallel has a `-k` option to keep the output order the same as the order of commands given.

**Comment by sublimino** [comment]
I like this idea - I've ruminated on using Bash-native job control but as it would rely on tracking PIDs as arguments to the `wait` command it'll get nasty quickly, and may have bugs with PID re-use for sufficiently complex or subshell/spawn-heavy test suites.

I also don't mind the dependency on GNU parallel as it's what I prefer personally - better control than `xargs`, native support for Bash functions, more configurable invocation. 

Other parallel test runners (specifically [ava](https://github.com/avajs/ava/)) [do not guarantee ordering](https://github.com/avajs/ava/issues/1532#issuecomment-331714660) for parallelised tests and I suggest we do the same (and that guaranteeing anything here may spiral complexity).

What do you think @mbland @jasonkarns?

**Comment by cyphar** [comment]
> I also don't mind the dependency on GNU parallel as it's what I prefer personally - better control than `xargs`, native support for Bash functions, more configurable invocation.

Yeah, I agree -- not to mention that there are several mostly-compatible versions of GNU parallel if the license terms are unacceptable. Worst case, we could add support for them if there's some incompatibilities.

> Other parallel test runners (specifically [ava](https://github.com/avajs/ava/)) [do not guarantee ordering](https://github.com/avajs/ava/issues/1532#issuecomment-331714660) for parallelised tests and I suggest we do the same (and that guaranteeing anything here may spiral complexity).

I agree. Right now this patch doesn't guarantee that the order of execution of the tests -- rather just that the output is ordered (since GNU parallel gives us `-k` for free I thought I might as well use it).

**Comment by jasonkarns** [comment]
I'm excited about this.



> the dependency on GNU parallel

No problem since the fallback (running serially) still "works". And we can add other alternatives if desired later.

> ordering...

A couple thoughts.

1. keeping the _output_ ordered is worthwhile for debugging, if it's not too hard
2. making the test invocation order random would be worthwhile. (randomized test runs helps prevent accidental inter-test dependencies. usually a mistake with setup/teardown) So I'd like to eventually _force_ random order, so long as we can print the randomization seed (And also take the seed as input in order to reproduce a particular order if an inter-test ordering bug is found.)

**Comment by cdevinesr** [comment]
Personally, I like the idea of parallel running of tests, and random order by default.  I do think, despite the concern for inter-test dependencies, that random order should be able to be toggled off.  I personally know of some use cases that for admittedly silly reasons depend on that structured order of execution, so just completely removing the ability to do it entirely might be overkill, even if random is certainly a better default.

**Comment by cyphar** [comment]
Random order would be fairly simple (assuming that you want the same order to be used for output), we just need to have a `sort -R`. *However* if we want to have random ordering, we should include the ability to re-use a given random ordering. Luckily there's a fairly easy way to do that (`sort --random-source=FILE -R`), and all we have to do is output the random source we use as diagnostic information (a line starting with `#`). Though, I'm not sure if macOS's `sort` has those features.

#172 is the current implementation, and it works okay as far as I can tell.

**Comment by sublimino** [comment]
I'm in favour of merging #172 in its current form, and then dealing with deterministic shuffling in a separate issue.

- [as mentioned by @jasonkarns](https://github.com/bats-core/bats-core/issues/171#issuecomment-460734802) it's non-critical as if a sequentially passing test suite fails in `--jobs n>1` mode as users still have the default sequential behaviour - and they can at least use test selectors to run pairs of tests in parallel to fix inter-test dependencies
  - arguably incrementally adding tests to parallel execution is a more certain debugging experience than fixing one special ordering case at a time (as represented by the seed value)
- inconsistencies between Linux and OSX make deterministic shuffling non-trivial and introduce another dependency (we strive hard to avoid anything non-Bash, although GNU parallel is a necessary exception IMHO, and not mandatory for core Bats behaviour)
- other popular parallel test runners do not currently support seeded randomisation/shuffling (although it's a desired feature and [open issues exist](https://github.com/avajs/ava/issues/595))

So this PR could be merged with a docs note to say:

> Ordering of parallised tests is not guaranteed, so this mode may break suites with dependencies between tests. When enabling `--jobs` for the first time be sure to re-run bats multiple times to identify any inter-test dependencies or non-deterministic test behaviour.

WDYT @jasonkarns @cdevinesr @cyphar?

**Comment by cyphar** [comment]
Sounds good to me -- where in the docs should I add that information? In README or somewhere else?

**Comment by andrewrech** [comment]
Hello all,

I am a bats user who currently parallelizes at the test level for HPC/genomics uses, and I have followed this discussion with interest.

I agree that, for me, non-deterministic ordering that could uncover test dependencies would be a feature, not a deficiency.

Thank you for your effort developing this tool, I look forward to this PR being merged.

Sincerely,

Andrew

**Comment by sublimino** [comment]
@cyphar just the README is fine if we all agree on the wording 👍

**Comment by cyphar** [comment]
I just figured out how to implement `--shuffle`. Here is the patch, it's quite trivial. And it let's you re-do a deterministic run (by setting `BATS_SHUFFLE_SEED`). Oh, and it only requires `awk` and `shuf`.

```patch
From f79d8c1fa9cdb6cc0de105f572edb420969afc5f Mon Sep 17 00:00:00 2001
From: Aleksa Sarai <cyphar@cyphar.com>
Date: Mon, 11 Feb 2019 02:18:16 +1100
Subject: [PATCH] bats: add support for deteriministic --shuffle

There are many cases where tests accidentially depend on one another,
and bats users might want to ensure their test suite does not have this
behaviour. While --jobs gives you some randomness in the execution,
--shuffle allows for deterministic ordering in order to re-test a
particular case (by setting BATS_SHUFFLE_SEED) and also shuffles all
tests entirely. And it only requires GNU coreutils (awk and shuf).

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>
---
 README.md                         |  3 ++-
 libexec/bats-core/bats            |  6 +++++-
 libexec/bats-core/bats-exec-suite | 36 +++++++++++++++++++++++++++++--
 3 files changed, 41 insertions(+), 4 deletions(-)

diff --git a/README.md b/README.md
index b9e0444893e0..e15fd7dfe0ee 100644
--- a/README.md
+++ b/README.md
@@ -179,7 +179,7 @@ supports:
 
 ```
 Bats x.y.z
-Usage: bats [-cr] [-f <regex>] [-j <jobs>] [-p | -t] <test>...
+Usage: bats [-crs] [-f <regex>] [-j <jobs>] [-p | -t] <test>...
        bats [-h | -v]
 
   <test> is the path to a Bats test file, or the path to a directory
@@ -191,6 +191,7 @@ Usage: bats [-cr] [-f <regex>] [-j <jobs>] [-p | -t] <test>...
   -j, --jobs       Number of parallel jobs to run (requires GNU parallel)
   -p, --pretty     Show results in pretty format (default for terminals)
   -r, --recursive  Include tests in subdirectories
+  -s, --shuffle    Shuffle the order of tests (using BATS_SHUFFLE_SEED, if set)
   -t, --tap        Show results in TAP format
   -v, --version    Display the version number
 
diff --git a/libexec/bats-core/bats b/libexec/bats-core/bats
index 03c1316f08b4..2a0799b20964 100755
--- a/libexec/bats-core/bats
+++ b/libexec/bats-core/bats
@@ -20,7 +20,7 @@ usage() {
   while IFS= read -r line; do
     printf '%s\n' "$line"
   done <<END_OF_HELP_TEXT
-Usage: $cmd [-cr] [-f <regex>] [-j <jobs>] [-p | -t] <test>...
+Usage: $cmd [-crs] [-f <regex>] [-j <jobs>] [-p | -t] <test>...
        $cmd [-h | -v]
 
   <test> is the path to a Bats test file, or the path to a directory
@@ -32,6 +32,7 @@ Usage: $cmd [-cr] [-f <regex>] [-j <jobs>] [-p | -t] <test>...
   -j, --jobs       Number of parallel jobs to run (requires GNU parallel)
   -p, --pretty     Show results in pretty format (default for terminals)
   -r, --recursive  Include tests in subdirectories
+  -s, --shuffle    Shuffle the order of tests (using BATS_SHUFFLE_SEED, if set)
   -t, --tap        Show results in TAP format
   -v, --version    Display the version number
 
@@ -126,6 +127,9 @@ while [[ "$#" -ne 0 ]]; do
   -p|--pretty)
     pretty=1
     ;;
+  -s|--shuffle)
+    flags+=('-s')
+    ;;
   -*)
     abort "Bad command line option '$1'"
     ;;
diff --git a/libexec/bats-core/bats-exec-suite b/libexec/bats-core/bats-exec-suite
index eb30d74ba00f..05d5d54802bc 100755
--- a/libexec/bats-core/bats-exec-suite
+++ b/libexec/bats-core/bats-exec-suite
@@ -1,7 +1,25 @@
 #!/usr/bin/env bash
 set -e
 
+random_stream() {
+  awk="$(type -p gawk awk | head -1)"
+  yes | "$awk" -v SEED="$1" \
+           'BEGIN { srand(SEED) } { printf("%c", 255 * rand()) }'
+}
+
+random_number() {
+  random_stream "$RANDOM$RANDOM" | tr -cd '0-9' | fold -w 16 | head -1
+}
+
+random_shuffle() {
+  shuf="$(type -p gshuf shuf | head -1)"
+  shuf --random-source <(random_stream "$BATS_SHUFFLE_SEED")
+}
+
+BATS_SHUFFLE_SEED="${BATS_SHUFFLE_SEED:-$(random_number)}"
+
 count_only_flag=''
+shuffle_flag=''
 extended_syntax_flag=''
 filter=''
 num_jobs=1
@@ -22,6 +40,9 @@ while [[ "$#" -ne 0 ]]; do
     shift
     num_jobs="$1"
     ;;
+  -s)
+    shuffle_flag=1
+    ;;
   -x)
     extended_syntax_flag='-x'
     flags+=('-x')
@@ -84,11 +105,22 @@ if [[ "${#all_tests[@]}" == 0 ]]; then
   exit
 fi
 
+shuffle_cmd=cat
+if [[ -n "$shuffle_flag" ]]; then
+  printf '# bats is running with --shuffle, the test suite order is randomised!\n'
+  printf '# In order to reproduce this run, set BATS_SHUFFLE_SEED=%s in your environment.\n' "$BATS_SHUFFLE_SEED"
+  shuffle_cmd=random_shuffle
+fi
+
+output_tests() {
+  printf '%s\n' "${all_tests[@]}" | grep -v '^$' | "$shuffle_cmd"
+}
+
 if [[ "$num_jobs" != 1 ]]; then
   # Only use GNU parallel when we want parallel execution -- there is a small
   # amount of overhead using it over a simple loop in the serial case.
   set -o pipefail
-  printf '%s\n' "${all_tests[@]}" | grep -v '^$' | \
+  output_tests | \
     parallel -qk -j "$num_jobs" --colsep="\t" -- bats-exec-test "${flags[@]}" '{1}' '{2}' '{#}' || status=1
 else
   # Just do it serially.
@@ -96,6 +128,6 @@ else
   while IFS=$'\t' read -r filename test_name; do
     bats-exec-test "${flags[@]}" "$filename" "$test_name" "$test_number" || status=1
     ((++test_number))
-  done < <(printf '%s\n' "${all_tests[@]}" | grep -v '^$')
+  done < <(output_tests)
 fi
 exit "$status"
-- 
2.20.1

```

**Comment by cyphar** [comment]
#172 should be ready to merge.

**Comment by sublimino** [comment]
Parallel tests pass (in half the time!) against my projects with Bats suites.

Anybody else like to verify before we merge please?

n.b. for changes like this: automating the testing of the projects that we know depend on Bats would be tremendously useful

**Comment by chavafg** [comment]
Hi, 

I'v just tested this PR on my project's [test suite](https://github.com/kata-containers/tests/tree/master/integration/kubernetes) and works pretty well, reduces the time by more than a half.

Thanks for implementing this, looking forward to see it merged.

**Comment by patthiel** [comment]
Wow, and here i was investigating using GNU `parallel` to do exactly this. Thanks all involved in this effort! Excited to use this.
